packages/anchor-listener/src/loader.ts

```diff
@@ -0,0 +1,48 @@
+import { getCidFromAnchorEventLog } from '@ceramicnetwork/anchor-contract'
+import type { Block, BlockTag, Filter, Provider } from '@ethersproject/providers'
+import type { CID } from 'multiformats/cid'
+import {
+  type Observable,
+  type OperatorFunction,
+  type RetryConfig,
+  firstValueFrom,
+  from,
+  map,
+  mergeMap,
+  pipe,
+  retry,
+} from 'rxjs'
+
+export function createBlockLoader(
+  provider: Provider,
+  block: BlockTag,
+  retryConfig: RetryConfig = { count: 3 }
+): Observable<Block> {
+  return from(provider.getBlock(block)).pipe(retry(retryConfig))
+}
+
+export async function loadBlock(
+  provider: Provider,
+  block: BlockTag,
+  retryConfig?: RetryConfig
+): Promise<Block> {
+  return await firstValueFrom(createBlockLoader(provider, block, retryConfig))
+}
+
+export function mapLoadBlock(
+  provider: Provider,
+  retryConfig: RetryConfig = { count: 3 }
+): OperatorFunction<number, Block> {
+  return pipe(mergeMap((blockNumber) => createBlockLoader(provider, blockNumber, retryConfig)))
+}
+
+export function createRootCIDsLoader(
+  provider: Provider,
+  filter: Filter,
+  retryConfig: RetryConfig = { count: 3 }
+): Observable<Array<CID>> {
+  return from(provider.getLogs(filter)).pipe(
```

Comment:
> Keep in mind that many backends will discard old events, and that requests which are too broad may get dropped as they require too many resources to execute the query.
https://docs.ethers.io/v5/api/providers/provider/#Provider-getLogs

The above is found on the ethers documentation. I wonder if the batching(? this may not be the correct word) should be handled here or by the user of this package

Reply:
Yes I was wondering about that too.
I think that can be handled here based on the `fromBlock` and `toBlock` parameters of the `filter` argument and possible error response, but it depends if the consumer wants to handle these errors directly or not I guess?
A simple logic could be to only try to load half the number of blocks on error, until loading is successful?

Reply:
> A simple logic could be to only try to load half the number of blocks on error, until loading is successful?
I really like that! We can add functionality in the future if needed

Reply:
@stephhuynh18 I changed the type of the event emitted from the listener to https://github.com/ceramicnetwork/js-ceramic/pull/2599/files#diff-8edd94f3b92ba95835a94cf05acb3a4aeacaad9480c03c81346d669230316d14R32 does that make sense please?
The idea is that if the `reorganized` flag is `true`, the consumer can handle what it wants to do based on the `block` and `expectedParentHash` values, what do you think please?